tpu/templates/tpu_grl_estimator/trainer.py

```diff
@@ -0,0 +1,308 @@
+
+    source_tensor = tf.constant(source, dtype=tf.float32)
+    labels_tensor = tf.constant(labels, dtype=tf.int32)
+    target_tensor = tf.constant(target, dtype=tf.float32)
+
+    # shuffle source and target separately
+    source_labels_dataset = tf.data.Dataset.from_tensor_slices((source_tensor, labels_tensor)).repeat().shuffle(32)
+    target_dataset = tf.data.Dataset.from_tensor_slices(target_tensor).repeat().shuffle(32)
+
+    # zip them together to set shapes
+    dataset = tf.data.Dataset.zip((source_labels_dataset, target_dataset))
+
+    # TPUEstimator passes params when calling input_fn
+    batch_size = params.get('batch_size', 16)
+
+    dataset = dataset.batch(batch_size, drop_remainder=True)
+
+    # TPUs need to know all dimensions when the graph is built
+    # Datasets know the batch size only when the graph is run
+    def set_shapes_and_format(source_labels, target):
+        source, labels = source_labels
+
+        source_shape = source.get_shape().merge_with([batch_size, None])
+        labels_shape = labels.get_shape().merge_with([batch_size])
+        target_shape = target.get_shape().merge_with([batch_size, None])
+
+        source.set_shape(source_shape)
+        labels.set_shape(labels_shape)
+        target.set_shape(target_shape)
+
+        # Also format the dataset with a dict for features
+        features = {'source': source, 'target': target}
+
+        return features, labels
+
+    dataset = dataset.map(set_shapes_and_format)
+    dataset = dataset.prefetch(tf.contrib.data.AUTOTUNE)
+
+    return dataset
+
+
+def main(args):
+    # pass the args as params so the model_fn can use
+    # the TPU specific args
+    params = vars(args)
+
+    if args.use_tpu:
+        # additional configs required for using TPUs
+        tpu_cluster_resolver = tf.contrib.cluster_resolver.TPUClusterResolver(args.tpu)
+        tpu_config = tf.contrib.tpu.TPUConfig(
+            num_shards=8, # using Cloud TPU v2-8
+            iterations_per_loop=args.save_checkpoints_steps)
+
+        # use the TPU version of RunConfig
+        config = tf.contrib.tpu.RunConfig(
+            cluster=tpu_cluster_resolver,
+            model_dir=args.model_dir,
+            tpu_config=tpu_config,
+            save_checkpoints_steps=args.save_checkpoints_steps,
+            save_summary_steps=100)
+
+        # TPUEstimator
+        estimator = tf.contrib.tpu.TPUEstimator(
+            model_fn=model_fn,
+            config=config,
+            params=params,
+            train_batch_size=args.train_batch_size,
+            eval_batch_size=32, # FIXME
+            export_to_tpu=False)
+    else:
+        config = tf.estimator.RunConfig(model_dir=args.model_dir)
+
+        estimator = tf.estimator.Estimator(
+            model_fn,
+            config=config,
+            params=params)
+
+    estimator.train(train_input_fn, max_steps=args.max_steps)
+
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser()
+
+    parser.add_argument(
```

Comment:
Should we add help to args, explaining the meaning of args, (Even though some are self-explanatory)